### PR TITLE
Edit permissions

### PR DESCRIPTION
**Pull request by chetanyakan**
#### Summary

- Add file edit permissions setting in system console.
- If the feature is enabled, then file can be edited by the owner by default and the owner can share the file with the entire channel.

#### Screenshots

Peek 2021-06-24 20-10

### CODE REVIEW

File: webapp/src/components/file_preview_component.tsx [hunk 1]
```diff
@@ -13,15 +17,28 @@ type Props = {
 const FilePreviewComponent: FC<Props> = ({fileInfo}: Props) => {
     const [loading, setLoading] = useState(true);
     const [editable, setEditable] = useState(false);
-    const enableEditing = useCallback(() => setEditable(true), []);
+    const enableEditing = useCallback(() => {
+        setEditable(true);
+    }, []);
+
+    const getIsCurrentUserFileOwner = makeGetIsCurrentUserFileOwner();
+    const getCollaboraFilePermissions = makeGetCollaboraFilePermissions();
```
Comment (manojmalik20) [inline webapp/src/components/file_preview_component.tsx]:
```suggestion
    const getIsCurrentUserFileOwner = useMemo(makeGetIsCurrentUserFileOwner, []);
    const getCollaboraFilePermissions = useMemo(makeGetCollaboraFilePermissions, []);
```

File: webapp/src/components/file_preview_modal.tsx [hunk 2]
```diff
@@ -1,35 +1,80 @@
...
-    const dispatch = useDispatch();
+    const dispatch: ThunkDispatch<GlobalState, undefined, AnyAction> = useDispatch();
     const {visible, fileInfo}: FilePreviewModalSelector = useSelector(filePreviewModal);
     const [editable, setEditable] = useState(false);
     const toggleEditing = useCallback(() => {
         setEditable((prevState) => !prevState);
     }, [setEditable]);
 
+    const getIsCurrentUserFileOwner = makeGetIsCurrentUserFileOwner();
+    const getCollaboraFilePermissions = makeGetCollaboraFilePermissions();
```
Comment (manojmalik20) [inline webapp/src/components/file_preview_modal.tsx]:
```suggestion
    const getIsCurrentUserFileOwner = useMemo(makeGetIsCurrentUserFileOwner, []);
    const getCollaboraFilePermissions = useMemo(makeGetCollaboraFilePermissions, []);
```

**Review by manojmalik20 — APPROVED** [review]

File: webapp/src/components/file_preview_modal.tsx [hunk 3]
```diff
@@ -1,35 +1,80 @@
...
+    const isCurrentUserOwner = useSelector((state: GlobalState) => getIsCurrentUserFileOwner(state, fileInfo));
+    const filePermission = useSelector((state: GlobalState) => getCollaboraFilePermissions(state, fileInfo));
+    const editPermissionsFeatureEnabled = useSelector(collaboraFileEditPermissionsEnabled);
+    const showEditPermissionChangeOption = editPermissionsFeatureEnabled && isCurrentUserOwner;
+
+    const [canChannelEdit, setCanChannelEdit] = useState(false);
+    const toggleCanChannelEdit = async () => {
+        await setCanChannelEdit((prevState: boolean) => !prevState);
```
Comment (hrishavkmr) [inline webapp/src/components/file_preview_modal.tsx:46]: why this await here?

**Review by manojmalik20 — APPROVED** [review]

File: webapp/src/components/file_preview_modal.tsx (hunk 3 above)
Comment (chetanyakan) [reply webapp/src/components/file_preview_modal.tsx:46]: This is required to correctly use the state in the API call below.

**Pull request merged**